File: LeetCode/94_BinartTreeInorderTraversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# 2021 12.07

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # # v1 Recursive inorder traversal
        # if root == None:
        #     return []
        # ret=[root.val]
        # if root.left:
        #     ret=self.inorderTraversal(root.left)+ret
        # if root.right:
        #     ret=ret+self.inorderTraversal(root.right)
        # return ret

        # # v1.1 recursive oneliner
        # return self.inorderTraversal(root.left)+[root.val]+self.inorderTraversal(root.right) if root else []

        # v2 copied iterative, non-recursive using stack
        ret =[]
        stack=[]
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root=stack.pop()
            ret.append(root.val)
            root=root.right
        return ret

File: LeetCode/94_BinartTreeInorderTraversal.java
// https://leetcode.com/problems/binary-tree-inorder-traversal/solutions/

// 2023年09月01日 10:15:23
// iterative best practise
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        Stack<TreeNode> s=new Stack<TreeNode>();
        List<Integer> ret=new ArrayList<Integer>();
        while(root!=null || !s.isEmpty()){
            while(root!=null){
                s.add(root);
                root=root.left;
            }
            root=s.pop();
            ret.add(root.val);
            root=root.right;
        }
        return ret;
    }
}

// recursive
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ret=new ArrayList<Integer>();
        if(root!=null){
            ret.addAll(inorderTraversal(root.left));
            ret.add(root.val);
            ret.addAll(inorderTraversal(root.right));
        }
        return ret;
    }
}
